SparseGrids/tsgRuleWavelet.cpp

```diff
@@ -295,35 +295,53 @@ double RuleWavelet::getWeight(int point) const{
     return 0.0;
 }
 
+template<int mode>
 double RuleWavelet::eval(int point, double x) const{
-    // Evaluates a wavelet designated by point at coordinate x.
+    // Evaluates or differentiates a wavelet designated by point at coordinate x.
     if(order == 1){
         // Level 0
         if (point < 3){
-            double w = 1.0 - std::abs(x - getNode(point));
-            return (w < 0.0) ? 0.0 : w;
+            double node = getNode(point);
+            if (mode == 0) {
+                double w = 1.0 - std::abs(x - node);
+                return (w < 0.0) ? 0.0 : w;
+            }else{
+                // Preserve the symmetry of derivatives as much as possible.
+                if (point == 0) return (x < 0.0) ? 1.0 : -1.0;
+                else if (point == 1) return (x < 0.0) ? -1.0 : 0.0;
+                else return (x < 0.0) ? 0.0 : 1.0;
+            }
         }
         // Level 1+
-        return eval_linear(point, x);
+        return eval_linear<mode>(point, x);
     }
     else if(order == 3){
-        return eval_cubic(point, x);
+        return eval_cubic<mode>(point, x);
     }
     return 0.;
 }
+template double RuleWavelet::eval<0>(int point, double x) const;
+template double RuleWavelet::eval<1>(int point, double x) const;
 
+template<int mode>
 inline double RuleWavelet::eval_cubic(int point, double x) const{
+    // Helps stabilize numerical errors.
+    if (point == 0 and x == 0.0 and mode == 1) return 0.0;
+
     // Evaluates a third order wavelet at a given point x.
+    double sgn = 1.0;
     if (point < 5){ // Scaling functions
         if (point == 2){ // Reflect across y-axis
             point = 1;
+            sgn = -1.0;
             x = -x;
         }else if(point == 4){
             point = 3;
+            sgn = -1.0;
             x = -x;
         }
         const double *phi = &(data[1][((point+1)/2) * num_data_points]);
-        return interpolate(phi, x);
+        return (mode == 0 ? 1.0 : sgn) * interpolate<mode>(phi, x);
```

Comment:
In `mode == 0` this still multiplies by 1. It will probably get optimized out, but still, an explicit way is:
```
return (mode == 0 ? interpolate<mode>(phi, x) : sgn * interpolate<mode>(phi, x));
```
Note that this will never evaluate `interpolate` twice regardless of the value of `mode`.

Reply:
Changed.